storage: Generate raft snapshots asynchronously

bdarnell opened the pull request:
Blocking the processRaft goroutine for too long is problematic. In
extreme cases it can cause heartbeats to be missed and new elections to
start (a major cause of #5970). This commit moves the work of snapshot
generation to an asynchronous goroutine.

Fixes #6204.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6253)

<!-- Reviewable:end -->

bdarnell committed:
storage: Generate raft snapshots asynchronously

Blocking the processRaft goroutine for too long is problematic. In
extreme cases it can cause heartbeats to be missed and new elections to
start (a major cause of #5970). This commit moves the work of snapshot
generation to an asynchronous goroutine.

Fixes #6204.

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 234 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-KGBU7d8bFKQ-tWwW-Q5:-KGBU7d8bFKQ-tWwW-Q6:-383186197) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e2cd8e21b957b8192eb734ef02675ad9cc60a3/storage/replica.go#L234)):*
Doesn't appear to be set back to `nil` after a snapshot generation completes. There seems little reason to do so.

---

*[storage/replica_raftstorage.go, line 384 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-KGBSBVojx8wxIfdP3fg:-KGBSBVojx8wxIfdP3fh:-2047957191) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e2cd8e21b957b8192eb734ef02675ad9cc60a3/storage/replica_raftstorage.go#L384)):*
why not move `snap` into the closure? Then you don't need the `failed to start task` part below and you also don't shadow `snap` as you do now.

---

*[storage/replica_raftstorage.go, line 398 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-KGBZdiSe4HyqPe5ckIv:-KGBZdiSe4HyqPe5ckIw:1757057394) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e2cd8e21b957b8192eb734ef02675ad9cc60a3/storage/replica_raftstorage.go#L398)):*
The comment is somewhat misleading because we don't really know what Raft decided, only that it hasn't been back for a while.

---

*[storage/replica_raftstorage.go, line 510 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-KGBStMRGkxIWK2cYUKm:-KGBStMRGkxIWK2cYUKn:483371824) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e2cd8e21b957b8192eb734ef02675ad9cc60a3/storage/replica_raftstorage.go#L510)):*
Looks like testing-only code. Can it be moved to `*_test.go`?

---

*[storage/replica_test.go, line 5198 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-KGBYNd02WNXFlvqL3iL:-KGBTYtvxjKdI5lr6SQo:149698321) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e2cd8e21b957b8192eb734ef02675ad9cc60a3/storage/replica_test.go#L5198)):*
Awkward sleep, but I don't see a good way to get rid of it.

---

*[storage/replica_test.go, line 5212 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-KGBZHzaCqsg99KnMSi7:-KGBZHzaCqsg99KnMSi8:-1079228097) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e2cd8e21b957b8192eb734ef02675ad9cc60a3/storage/replica_test.go#L5212)):*
In anticipation of CircleCI, could you make this ridiculously high?

---

*[storage/store.go, line 381 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-KGBZpUEBnXyu1nr9G_A:-KGBZpUEBnXyu1nr9G_B:-823403634) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e2cd8e21b957b8192eb734ef02675ad9cc60a3/storage/store.go#L381)):*
Add that it waits only when initiating work (i.e. when checking on an existing unit of work, no time will be spent waiting regardless of this setting).

---

*[storage/store.go, line 386 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-KGBZTwhZVC8XKMGRNhH:-KGBZTwhZVC8XKMGRNhI:-2072365165) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e2cd8e21b957b8192eb734ef02675ad9cc60a3/storage/store.go#L386)):*
Explain that it excludes the actual time spent creating the snapshot.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-:-KGBZztW52bJxRtpx_EB:1417088721)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 234 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-KGBU7d8bFKQ-tWwW-Q5:-KGD-9XGIdkkalPHxQdh:-1890285216) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e2cd8e21b957b8192eb734ef02675ad9cc60a3/storage/replica.go#L234)):*
Yeah, I originally reset it back to nil after reading from it, but that seemed error prone so I took all of those out to make sure it worked. I'll revise the comment.

---

*[storage/replica_raftstorage.go, line 384 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-KGBSBVojx8wxIfdP3fg:-KGD4oMsUFYwyuwujBN-:1529700183) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e2cd8e21b957b8192eb734ef02675ad9cc60a3/storage/replica_raftstorage.go#L384)):*
I wanted to grab the snapshot outside so it would be synchronized WRT other database operations but I guess that doesn't really matter for snapshots. Moved it into the closure.

---

*[storage/replica_raftstorage.go, line 398 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-KGBZdiSe4HyqPe5ckIv:-KGD5LA04GuUuiMDB0_6:1582050969) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e2cd8e21b957b8192eb734ef02675ad9cc60a3/storage/replica_raftstorage.go#L398)):*
Updated comment.

---

*[storage/replica_raftstorage.go, line 510 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-KGBStMRGkxIWK2cYUKm:-KGD5MJJ_iVuGnlmEahX:-1263897661) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e2cd8e21b957b8192eb734ef02675ad9cc60a3/storage/replica_raftstorage.go#L510)):*
Moved to `helpers_test.go`

---

*[storage/replica_test.go, line 5212 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-KGBZHzaCqsg99KnMSi7:-KGD5tjB2DH2eJs5sMlf:-1519201204) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e2cd8e21b957b8192eb734ef02675ad9cc60a3/storage/replica_test.go#L5212)):*
I think `time.Second` is already ridiculously high; it should only take a few milliseconds.

---

*[storage/store.go, line 381 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-KGBZpUEBnXyu1nr9G_A:-KGD8BnEaXKKc65oX9IG:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e2cd8e21b957b8192eb734ef02675ad9cc60a3/storage/store.go#L381)):*
Done.

---

*[storage/store.go, line 386 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-KGBZTwhZVC8XKMGRNhH:-KGD8CC0r8LX5rTz0FsE:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e2cd8e21b957b8192eb734ef02675ad9cc60a3/storage/store.go#L386)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (again)

---

Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/replica_test.go, line 5212 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-KGBZHzaCqsg99KnMSi7:-KGDKjzUMuJ8VxL0uobY:1718244326) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e2cd8e21b957b8192eb734ef02675ad9cc60a3/storage/replica_test.go#L5212)):*
I've seen log line timestamps jump by >1s, but I suppose we'll see it if it happens.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6253#-:-KGDKvUTavlVHLbFLbF0:-1562000222)*
<!-- Sent from Reviewable.io -->
